File: heap/19_sort_characters_by_frequency.cpp
#include <bits/stdc++.h>
using namespace std;

/*
Problem Description:
Given a string s, sort characters in decreasing order by frequency.

Sample Input/Output:
1) s="tree" -> "eetr" or "eert"
2) s="cccaaa" -> "aaaccc" or "cccaaa"
3) s="Aabb" -> "bbAa" or "bbaA"

Code Explanation:
Count character frequency and push {frequency, character} into max-heap.
Build output by popping highest frequency first.
*/

string frequencySort(string s) {
    unordered_map<char, int> freq;
    for (char c : s) freq[c]++;

    priority_queue<pair<int, char>> maxHeap;
    for (auto &entry : freq) maxHeap.push({entry.second, entry.first});

    string result;
    while (!maxHeap.empty()) {
        pair<int, char> top = maxHeap.top();
        maxHeap.pop();
        int f = top.first;
        char c = top.second;
        result.append(f, c);
    }
    return result;
}

int main() {
    cout << "Output 1: " << frequencySort("tree") << "\n";
    cout << "Output 2: " << frequencySort("cccaaa") << "\n";
    cout << "Output 3: " << frequencySort("Aabb") << "\n";
    return 0;
}

File: heap/19_sort_characters_by_frequency.py
"""
PROBLEM: Sort Characters By Frequency (LeetCode 451)
DIFFICULTY: Medium
PATTERN: Heap - Max-Heap by character frequency

DESCRIPTION:
Given string s, sort it so characters appear in decreasing order of frequency.

EXAMPLE:
Input: s="tree" -> "eert" or "eetr"
Input: s="cccaaa" -> "cccaaa" or "aaaccc"
Input: s="Aabb" -> "bbAa" or "bbaA"

TIME COMPLEXITY: O(n log n)
SPACE COMPLEXITY: O(n)
"""

import heapq
from collections import Counter


class Solution:
    def frequencySort(self, s: str) -> str:
        freq = Counter(s)
        # Max-heap: negate frequency
        max_heap = [(-cnt, ch) for ch, cnt in freq.items()]
        heapq.heapify(max_heap)

        result = []
        while max_heap:
            cnt, ch = heapq.heappop(max_heap)
            result.append(ch * (-cnt))

        return "".join(result)


if __name__ == "__main__":
    sol = Solution()
    for s in ["tree", "cccaaa", "Aabb", "z"]:
        print(f"s={s!r} -> {sol.frequencySort(s)!r}")
